agg-ast/schema_derivation/src/schema_derivation.rs

```diff
@@ -208,6 +208,88 @@ impl DeriveSchema for Stage {
             Ok(state.result_set_schema.to_owned())
         }
 
+        fn group_derive_schema(group: &Group, state: &mut ResultSetState) -> Result<Schema> {
+            // group is a map of field namespace to expression. We can derive the schema for each expression
+            // and then union them together to get the resulting schema.
+            let mut keys = group
+                .aggregations
+                .iter()
+                .map(|(k, e)| {
+                    let field_schema = e.derive_schema(state)?;
+                    Ok((k.to_string(), field_schema))
+                })
+                .collect::<Result<BTreeMap<String, Schema>>>()?;
+            keys.insert("_id".to_string(), group.keys.derive_schema(state)?);
+            Ok(Schema::Document(Document {
+                required: keys.keys().cloned().collect(),
+                keys,
+                ..Default::default()
+            }))
+        }
+
+        // add_fields_derive_schema is near set_window_fields_derive_schema because they are necessarily
+        // identical in semantics: they both add fields to the schema incoming from the previous
+        // pipeline stage. The only difference is that add_fields is a map of expressions, while set_window_fields
+        // is a map of window functions.
+        fn add_fields_derive_schema(
```

Comment:
perhaps I'm missing a little nuance here but wouldn't we just directly mutate the result set schema and use the insert_required_... helper?
```
for (field, expression) in add_fields.iter() {
    let expr_schema = expression.derive_schema(state)?;
    let path = field
        .split(".")
        .map(|s| s.to_string())
        .collect::<Vec<String>>();
    insert_required_key_into_document(
        &mut state.result_set_schema,
        expr_schema,
        path,
        true,
    );
}
Ok(Schema::simplify(&state.result_set_schema.to_owned()))
```

Reply:
Nah, that seems like a good idea.